lib/moab/deposit_bag_validator.rb

```diff
@@ -0,0 +1,324 @@
+              end
+            else
+              result_array << single_error_hash(CHECKSUM_TYPE_UNRECOGNIZED, checksum_type: checksum_type, filename: child_pathname)
+            end
+          end
+        end
+      end
+      checksums_hash
+    end
+
+    # generate hash of checksums by file name for bag home dir files
+    def generate_tagmanifest_checksums_hash(types_to_generate)
+      # all names in the bag home dir except those starting with 'tagmanifest'
+      home_dir_pathnames = deposit_bag_pathname.children.reject { |file| file.basename.to_s.start_with?(TAGMANIFEST) }
+      hash_with_full_pathnames = generate_checksums_hash(home_dir_pathnames, types_to_generate)
+      # return hash keys as basenames only
+      hash_with_full_pathnames.map { |k, v| [Pathname.new(k).basename.to_s, v] }.to_h
+    end
+
+    # generate hash of checksums by file name for bag data dir files
+    def generate_payload_checksums(types_to_generate)
+      data_pathnames = deposit_bag_pathname.join(DATA_DIR_BASENAME).find
+      hash_with_full_pathnames = generate_checksums_hash(data_pathnames, types_to_generate)
+      # return hash keys beginning with 'data/'
+      hash_with_full_pathnames.map { |k, v| [Pathname.new(k).relative_path_from(deposit_bag_pathname).to_s, v] }.to_h
+    end
+
+    def generate_checksums_hash(pathnames, types_to_generate)
+      file_checksums_hash = {}
+      pathnames.each do |pathname|
+        file_checksums_hash[pathname.to_s] = generated_checksums(pathname, types_to_generate) if pathname.file?
+      end
+      file_checksums_hash
+    end
+
+    def generated_checksums(pathname, types_to_generate)
+      my_digester_hash = digester_hash(types_to_generate)
+      pathname.open('r') do |stream|
+        while (buffer = stream.read(8192))
+          my_digester_hash.each_value { |digest| digest.update(buffer) }
+        end
+      end
+      file_checksums = {}
+      my_digester_hash.each do |checksum_type, digest|
+        file_checksums[checksum_type] = digest.hexdigest
+      end
+      file_checksums
+    end
+
+    def digester_hash(types_to_generate=DEFAULT_CHECKSUM_TYPES)
+      types_to_generate.each_with_object({}) do |checksum_type, digester_hash|
+        case checksum_type
+        when :md5
+          digester_hash[checksum_type] = Digest::MD5.new
+        when :sha1
+          digester_hash[checksum_type] = Digest::SHA1.new
+        when :sha256
+          digester_hash[checksum_type] = Digest::SHA2.new(256)
+        when :sha384
+          digesters[checksum_type] = Digest::SHA2.new(384)
+        when :sha512
+          digesters[checksum_type] = Digest::SHA2.new(512)
+        else
+          result_array << single_error_hash(CHECKSUM_TYPE_UNRECOGNIZED, checksum_type: checksum_type, filename: nil)
+        end
+        digester_hash
+      end
+    end
+
+    def verify_manifest_checksums(manifest_type, manifests_checksum_hash, generated_checksum_hash)
```

Comment:
i see ``` verify_manifest_checksums``` is the combination of ```verify_manifests``` and ```manifest_diff```. I guess ```manifest_diff``` uses ``` FileFixity.new(file_id: file_id)``` and we are using an empty hash here. Just wanted some clarification

Reply:
```     def initialize(options=nil)
      @checksums=Hash.new
      options = {} if options.nil?
      options.each do |key,value|
        #instance_variable_set("@#{key}", value)
        send "#{key}=", value
      end
    end
```

So they are passing in ```file_id``` to this hash obj but we are leaving it empty?

Reply:
So in the original, they use this "FileFixity" structure as a mean of associating a file name with its checksums.  And there is a bunch of code about making FileFixity a class that can be used as a hash key ... but it never is.

When I unraveled what was going on, it appeared we needed something simpler:  a hash with the relative file name as the key, and a hash of checksums as the value.   No need for all the indirection.

We don't have to worry about file name clashes, because they can't happen with relative file names in the bag.  And we are only verifying the checksums that exist for the file, so it seemed expedient to simply work with those.

It may be that some comments would be helpful explaining the hashes.

It may also be that I am wrong in my assessment.

Reply:
(Perhaps it was a bad decision that the original ingest-robots didn't leverage the moab-versioning code for the bag validation, since they do leverage it for deposit-bag.  I have no idea why the existing production code is so inconsistent and strange.)